### PR TITLE
[libc++][RFC] Only include what is required by-version in the umbrella headers

### PR DESCRIPTION
**Pull request by philnik777**
This is a relatively low cost way of reducing the include sizes in older language modes compared to the effect. For example, in C++14 mode the include time of `<algorithm>` is reduced from 198ms to 127ms.

### CODE REVIEW

**Comment by philnik777** [comment]
@mordante @ldionne What are your thoughts on this?

**Review by mordante — COMMENTED** [review]
As mentioned earlier I had a patch on Phab with a slightly different approach. So yes in general I like the idea of reducing the size of the includes.

I wonder about the approach you take; your examples only show how to handle granularized headers. I wonder what do you propose for non-granularized headers?

For example, `<ostream>` includes `<print>` the `<print>` header has been introduced in C++23. This header has some includes not used in `<ostream>`.

Do you propose to keep the `<ostream>` header unchanged or do you want to guard the include of `<print>` with C++23?

**Comment by philnik777** [comment]
> As mentioned earlier I had a patch on Phab with a slightly different approach. So yes in general I like the idea of reducing the size of the includes.
> 
> I wonder about the approach you take; your examples only show how to handle granularized headers. I wonder what do you propose for non-granularized headers?
> 
> For example, `<ostream>` includes `<print>` the `<print>` header has been introduced in C++23. This header has some includes not used in `<ostream>`.
> 
> Do you propose to keep the `<ostream>` header unchanged or do you want to guard the include of `<print>` with C++23?

I propose to either keep them unchanged, or to granularize it if we think it's worth the reduced include size. IIUC the main problem we have is `<format>` and the different `formatter` specializations that have to be provided through different headers resulting in huge amounts of transitive includes. Since they can be implemented separately from the rest of the files, we can simply move them into new detail headers and include these detail headers conditionally in the umbrella header.
In essence, I want the only place where we conditionally include headers to be the umbrella headers.

**Comment by mordante** [comment]
> > As mentioned earlier I had a patch on Phab with a slightly different approach. So yes in general I like the idea of reducing the size of the includes.
> > I wonder about the approach you take; your examples only show how to handle granularized headers. I wonder what do you propose for non-granularized headers?
> > For example, `<ostream>` includes `<print>` the `<print>` header has been introduced in C++23. This header has some includes not used in `<ostream>`.
> > Do you propose to keep the `<ostream>` header unchanged or do you want to guard the include of `<print>` with C++23?
> 
> I propose to either keep them unchanged, or to granularize it if we think it's worth the reduced include size. IIUC the main problem we have is `<format>` and the different `formatter` specializations that have to be provided through different headers resulting in huge amounts of transitive includes. Since they can be implemented separately from the rest of the files, we can simply move them into new detail headers and include these detail headers conditionally in the umbrella header. In essence, I want the only place where we conditionally include headers to be the umbrella headers.

I'm not sure how feasible splitting-off format would be. For example the `formatter` for `vector` needs to know about `vector`. IMO that is something we can look into for a separate patch if we want to solve these includes.

I think the approach you suggest is simple and easy to maintain. I think we could do more, but that might have an different maintenance trade-off. Taking this approach does not block other optimizations. So I'm happy to move forward with this direction.

File: libcxx/include/algorithm [hunk 1]
```diff
@@ -1796,213 +1796,219 @@ template <class BidirectionalIterator, class Compare>
...
-#include <__algorithm/sort_heap.h>
-#include <__algorithm/stable_partition.h>
-#include <__algorithm/stable_sort.h>
-#include <__algorithm/swap_ranges.h>
-#include <__algorithm/transform.h>
-#include <__algorithm/unique.h>
-#include <__algorithm/unique_copy.h>
-#include <__algorithm/unwrap_iter.h>
-#include <__algorithm/upper_bound.h>
+#if 1 // C++03/11
```
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/include/algorithm:1799]: I agree here, this feels like a leftover to me. I'd rather include unconditionally.

**Review by ldionne — APPROVED** [review]
This direction LGTM. There are a few bugs as pointed out by Mark, but with those fixed I think this patch makes a lot of sense.

**Pull request merged**